File: javascript_version/binary_trees/tree_levels/tree_levels.js
// tree levels
// Write a function, 
// treeLevels, that takes in a pointer to the root of a binary tree. 
// The function should return a 2-Dimensional vector where each subvector represents a level of the tree.


class Node {
    constructor (val) {
        this.val = val;
        this.left = null;
        this.right = null;
    }
}

const treeLevels = (root) => {
    // todo
    if (root === null) {
        return [];
    }
    let levels = []
    return _treeLevels(root, levels, 0);
};


const _treeLevels = (root, levels, levelNum) => {
    // todo
    if (root === null) {
        return [];
    }

    if (levels.length === levelNum) {
        levels[levelNum] = [root.val];
    }
    else {
        levels[levelNum].push(root.val);
    }
    let left = _treeLevels(root.left, levels, levelNum + 1);
    let right = _treeLevels(root.right, levels, levelNum + 1);
    return levels;
};
// const treeLevelsBreadth = (root) => {
//     //      a
//     //    /   \
//     //   b     c
//     //  / \     \
//     // d   e     f

//     if (root === null) {
//         return [];
//     }
//     let result = [];
//     let queue = [{ node: root, level: 0 }];
//     let prevLvl = 0;
//     let sub = [];
//     while (queue.length !== 0) {
//         let nodeInfo = queue.shift();
//         let node = nodeInfo.node;
//         let nodelvl = nodeInfo.level;

//         if (prevLvl === nodelvl) {
//             sub.push(node.val)
//         }
//         else {
//             result.push(sub);
//             prevLvl = nodelvl
//             sub = [node.val];
//         }
//         if (node.left !== null) {
//             queue.push({ node: node.left, level: nodelvl + 1 });
//         }
//         if (node.right !== null) {
//             queue.push({ node: node.right, level: nodelvl + 1 });
//         }
//     }
//     result.push(sub);
//     return result;
// };
//optimized version
const treeLevelsBreadth = (root) => {
    //      a
    //    /   \
    //   b     c
    //  / \     \
    // d   e     f

    if (root === null) { return []; }
    let result = [];
    let queue = [{ node: root, level: 0 }];
    while (queue.length !== 0) {
        let { node, level } = queue.shift();
        if (result.length === level) {
            result[level] = [node.val];
        }
        else {
            result[level].push(node.val);
        }
        if (node.left !== null) {
            queue.push({ node: node.left, level: level + 1 });
        }
        if (node.right !== null) {
            queue.push({ node: node.right, level: level + 1 });
        }
    }
    return result;
};

const a = new Node("a");
const b = new Node("b");
const c = new Node("c");
const d = new Node("d");
const e = new Node("e");
const f = new Node("f");

a.left = b;
a.right = c;
b.left = d;
b.right = e;
c.right = f;

//      a
//    /   \
//   b     c
//  / \     \
// d   e     f

console.table(treeLevels(a)); // ->
console.table(treeLevelsBreadth(a)); // ->

// [
//   ['a'],
//   ['b', 'c'],
//   ['d', 'e', 'f']
// ]
// const a = new Node("a");
// const b = new Node("b");
// const c = new Node("c");
// const d = new Node("d");
// const e = new Node("e");
// const f = new Node("f");
const g = new Node("g");
const h = new Node("h");
const i = new Node("i");

a.left = b;
a.right = c;
b.left = d;
b.right = e;
c.right = f;
e.left = g;
e.right = h;
f.left = i;

//         a
//      /    \
//     b      c
//   /  \      \
//  d    e      f
//      / \    /
//     g  h   i

console.table(treeLevels(a)); // ->
console.table(treeLevelsBreadth(a)); // ->

// [
//   ['a'],
//   ['b', 'c'],
//   ['d', 'e', 'f'],
//   ['g', 'h', 'i']
// ]
const q = new Node("q");
const r = new Node("r");
const s = new Node("s");
const t = new Node("t");
const u = new Node("u");
const v = new Node("v");

q.left = r;
q.right = s;
r.right = t;
t.left = u;
u.right = v;

//      q
//    /   \
//   r     s
//    \
//     t
//    /
//   u
//  /
// v

console.table(treeLevels(q)); //->
console.table(treeLevelsBreadth(q)); //->

// [
//   ['q'],
//   ['r', 's'],
//   ['t'],
//   ['u'],
//   ['v']
// ]
console.table(treeLevels(null)); // -> []
console.table(treeLevelsBreadth(null)); // -> []

File: cpp_version/binary_trees/tree_levels/tree_levels.cpp
#include <algorithm>
#include <cmath>
#include <iostream>
#include <limits.h>
#include <math.h>
#include <queue>
#include <stack>
#include <string>
#include <tuple>
#include <unordered_map>
#include <unordered_set>
#include <vector>

// tree levels
// Write a function,
// treeLevels, that takes in a pointer to the root of a binary tree.
// The function should return a 2-Dimensional vector where each subvector represents a level of the tree.

class Node {
public:
    std::string val;
    Node *left;
    Node *right;

    Node(std::string initialVal) {
        val = initialVal;
        left = nullptr;
        right = nullptr;
    }
};

std::vector<std::vector<std::string>> treeLevels(Node *);
std::vector<std::vector<std::string>> _treeLevels(Node *, std::vector<std::vector<std::string>> &, int);

std::vector<std::vector<std::string>> treeLevelsBreadth(Node *);

void printVector(std::vector<std::vector<std::string>>);
std::vector<std::vector<std::string>> treeLevels(Node *root) {
    // todo
    std::vector<std::vector<std::string>> levels;

    return _treeLevels(root, levels, 0);
}

std::vector<std::vector<std::string>> _treeLevels(Node *root, std::vector<std::vector<std::string>> &levels, int levelNum) {
    // todo



    return levels;
}

std::vector<std::vector<std::string>> treeLevelsBreadth(Node *root) {
    // todo

    std::vector<std::vector<std::string>> result;


    return result;
}
void printVector(std::vector<std::vector<std::string>> vec) {
    for (auto x : vec) {
        std::cout << "{ " << std::endl;
        ;
        std::cout << "{ ";
        for (auto y : x) {

            std::cout << y << " , ";
        }
        std::cout << " }" << std::endl;
    }
    std::cout << " }" << std::endl;
}

int main() {
    Node a("a");
    Node b("b");
    Node c("c");
    Node d("d");
    Node e("e");
    Node f("f");

    a.left = &b;
    a.right = &c;
    b.left = &d;
    b.right = &e;
    c.right = &f;

    //      a
    //    /   \
//   b     c
    //  / \     \
// d   e     f

    printVector(treeLevels(&a));        // ->
    printVector(treeLevelsBreadth(&a)); // ->

    // [
    //   ["a"],
    //   ["b", "c"],
    //   ["d", "e", "f"]
    // ]
    // Node a("a");
    // Node b("b");
    // Node c("c");
    // Node d("d");
    // Node e("e");
    // Node f("f");
    Node g("g");
    Node h("h");
    Node i("i");

    a.left = &b;
    a.right = &c;
    b.left = &d;
    b.right = &e;
    c.right = &f;
    e.left = &g;
    e.right = &h;
    f.left = &i;

    //         a
    //      /    \
//     b      c
    //   /  \      \
//  d    e      f
    //      / \    /
    //     g  h   i

    printVector(treeLevels(&a));        // ->
    printVector(treeLevelsBreadth(&a)); // ->

    // [
    //   ["a"],
    //   ["b", "c"],
    //   ["d", "e", "f"],
    //   ["g", "h", "i"]
    // ]
    Node q("q");
    Node r("r");
    Node s("s");
    Node t("t");
    Node u("u");
    Node v("v");

    q.left = &r;
    q.right = &s;
    r.right = &t;
    t.left = &u;
    u.right = &v;

    //      q
    //    /   \
//   r     s
    //    \
//     t
    //    /
    //   u
    //  /
    // v

    printVector(treeLevels(&q));        //->
    printVector(treeLevelsBreadth(&q)); //->

    // [
    //   ["q"],
    //   ["r", "s"],
    //   ["t"],
    //   ["u"],
    //   ["v"]
    // ]
    printVector(treeLevels(nullptr));        // -> []
    printVector(treeLevelsBreadth(nullptr)); // -> []

    return 0;
}